dmoj/judge.py

```diff
@@ -174,108 +93,128 @@ def begin_grading(
         report=print,
         blocking=False,
     ):
-        self.current_submission_id = id
+        self.current_submission_id = submission_id
+        sub_judge_thread = threading.Thread(
+            target=self._run_grading_thread,
+            args=(submission_id, problem_id, language, source, time_limit, memory_limit, short_circuit, meta, report),
+            daemon=True,
+        )
+        sub_judge_thread.start()
 
-        def grading_cleanup_wrapper():
+        # Submission may have been killed already, but block if it hasn't been.
+        if blocking and sub_judge_thread is not None:
+            sub_judge_thread.join()
+
+    def _run_grading_thread(
+        self, submission_id, problem_id, language, source, time_limit, memory_limit, short_circuit, meta, report
+    ):
+        try:
             report(
                 ansi_style(
-                    'Start grading #ansi[%s](yellow)/#ansi[%s](green|bold) in %s...' % (problem_id, id, language)
+                    'Start grading #ansi[%s](yellow)/#ansi[%s](green|bold) in %s...'
+                    % (problem_id, submission_id, language)
                 )
             )
 
-            try:
-                problem = Problem(problem_id, time_limit, memory_limit, meta)
-                self._block_and_grade(problem, language, source, short_circuit, report=report)
-            except Exception:
-                self.log_internal_error()
+            self.current_judge_process = SubmissionJudgeProcess(
+                submission_id, problem_id, language, source, time_limit, memory_limit, short_circuit, meta
+            )
+
+            ipc_handler_dispatch = {
+                IPC_COMPILE_ERROR: self._ipc_compile_error,
+                IPC_COMPILE_MESSAGE: self._ipc_compile_message,
+                IPC_GRADING_BEGIN: self._ipc_grading_begin,
+                IPC_GRADING_END: self._ipc_grading_end,
+                IPC_GRADING_ABORTED: self._ipc_grading_aborted,
+                IPC_BATCH_BEGIN: self._ipc_batch_begin,
+                IPC_BATCH_END: self._ipc_batch_end,
+                IPC_RESULT: self._ipc_result,
+                IPC_UNHANDLED_EXCEPTION: self._ipc_unhandled_exception,
+            }
+
+            for ipc_msg in self.current_judge_process.run():
+                type, data = ipc_msg
+
+                try:
+                    handler_func = ipc_handler_dispatch[type]
+                except KeyError:
+                    raise RuntimeError("unexpected IPC message %s" % str(ipc_msg)) from None
+
+                handler_func(report, *(data or tuple()))
```

Comment:
```suggestion
                handler_func(report, *(data or ()))
```
`tuple()` is a dumb way of saying `()`.

Reply:
Done.